politeiad/backendv2/tstorebe/plugins/ticketvote/hooks.go

```diff
@@ -0,0 +1,370 @@
+	if linkBy == 0 {
+		// LinkBy as not been set
+		return nil
+	}
+	min := time.Now().Unix() + p.linkByPeriodMin
+	max := time.Now().Unix() + p.linkByPeriodMax
+	switch {
+	case linkBy < min:
+		e := fmt.Sprintf("linkby %v is less than min required of %v",
+			linkBy, min)
+		return backend.PluginError{
+			PluginID:     ticketvote.PluginID,
+			ErrorCode:    uint32(ticketvote.ErrorCodeLinkByInvalid),
+			ErrorContext: e,
+		}
+	case linkBy > max:
+		e := fmt.Sprintf("linkby %v is more than max allowed of %v",
+			linkBy, max)
+		return backend.PluginError{
+			PluginID:     ticketvote.PluginID,
+			ErrorCode:    uint32(ticketvote.ErrorCodeLinkByInvalid),
+			ErrorContext: e,
+		}
+	}
+	return nil
+}
+
+func (p *ticketVotePlugin) linkToVerify(linkTo string) error {
+	// LinkTo must be a public record
+	token, err := tokenDecode(linkTo)
+	if err != nil {
+		return backend.PluginError{
+			PluginID:     ticketvote.PluginID,
+			ErrorCode:    uint32(ticketvote.ErrorCodeLinkToInvalid),
+			ErrorContext: "invalid hex",
+		}
+	}
+	r, err := p.recordAbridged(token)
+	if err != nil {
+		if err == backend.ErrRecordNotFound {
+			return backend.PluginError{
+				PluginID:     ticketvote.PluginID,
+				ErrorCode:    uint32(ticketvote.ErrorCodeLinkToInvalid),
+				ErrorContext: "record not found",
+			}
+		}
+		return err
+	}
+	if r.RecordMetadata.Status != backend.StatusPublic {
+		e := fmt.Sprintf("record status is invalid: got %v, want %v",
+			backend.Statuses[r.RecordMetadata.Status],
+			backend.Statuses[backend.StatusPublic])
+		return backend.PluginError{
+			PluginID:     ticketvote.PluginID,
+			ErrorCode:    uint32(ticketvote.ErrorCodeLinkToInvalid),
+			ErrorContext: e,
+		}
+	}
+
+	// LinkTo must be a runoff vote parent record, i.e. has specified
+	// a LinkBy deadline.
+	parentVM, err := voteMetadataDecode(r.Files)
+	if err != nil {
+		return err
+	}
+	if parentVM == nil || parentVM.LinkBy == 0 {
+		return backend.PluginError{
+			PluginID:     ticketvote.PluginID,
+			ErrorCode:    uint32(ticketvote.ErrorCodeLinkToInvalid),
+			ErrorContext: "record not a runoff vote parent",
+		}
+	}
+
+	// The LinkBy deadline must not be expired
+	if time.Now().Unix() > parentVM.LinkBy {
+		return backend.PluginError{
+			PluginID:     ticketvote.PluginID,
+			ErrorCode:    uint32(ticketvote.ErrorCodeLinkToInvalid),
+			ErrorContext: "parent record linkby deadline has expired",
+		}
+	}
+
+	// The runoff vote parent record must have been approved in a vote.
+	vs, err := p.summaryByToken(token)
+	if err != nil {
+		return err
+	}
+	if vs.Status != ticketvote.VoteStatusApproved {
+		return backend.PluginError{
+			PluginID:     ticketvote.PluginID,
+			ErrorCode:    uint32(ticketvote.ErrorCodeLinkToInvalid),
+			ErrorContext: "parent record vote is not approved",
+		}
+	}
+
+	return nil
+}
+
+func (p *ticketVotePlugin) voteMetadataVerify(vm ticketvote.VoteMetadata) error {
+	switch {
+	case vm.LinkBy == 0 && vm.LinkTo == "":
```

Comment:
This switch needs doco.

Reply:
Done
https://github.com/lukebp/politeia/blob/tlog/politeiad/backendv2/tstorebe/plugins/ticketvote/hooks.go#L295